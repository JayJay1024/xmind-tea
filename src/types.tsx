export enum EnumBill { Out=0, In=1 };  // 支出收入
export enum EnumMonth { Jan, Fed, Mar, Apr, May, Jun, Jul, Aug, Seq, Oct, Nov, Dec };

// 账单
export type TypeBill = {
    time: Date,
    type: EnumBill,
    category?: string,
    amount: number
};

// 分类
export type TypeCategory = {
    id: string,
    name: string,
    type: EnumBill
}

// 收支
export type TypeAmountInOut = {
    in: number,
    out: number
}

export const ConstMonthsText = ['1月', '2月', '3月', '4月', '5月', '6月', '7月', '8月', '9月', '10月', '11月', '12月'];
