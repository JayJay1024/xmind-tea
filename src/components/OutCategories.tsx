import React from 'react';
import { Table, Divider } from 'antd';
import dayjs from 'dayjs';

import {
    EnumBill,
    EnumMonth,
    TypeBill,
    TypeCategory,
    ConstMonthsText
} from '../types';

type TypeProp = {
    dataBill: TypeBill[],
    dataCategory: TypeCategory[],
    monthIndex: EnumMonth,
}

const OutCategories: React.FunctionComponent<TypeProp> = (props): JSX.Element => {
    const outCategoryColumns = [{
        title: '账单类型',
        dataIndex: 'category',
        key: 'category',
        align: 'center' as any,
    }, {
        title: '支出金额统计(￥)',
        dataIndex: 'amount',
        key: 'amount',
        align: 'center' as any,
        sorter: (a: any, b: any) => a.amount - b.amount,
        defaultSortOrder: 'descend' as any
    }];

    const checkOutCategoryData = (): Array<object> => {  // 分类统计支出
        let mapCategoryAmount = new Map<string, number>();
        const dataBillForMonth = props.dataBill.filter(v => Number(dayjs(v.time).month()) === Number(props.monthIndex));  // 选择月份后的账单数据

        // => Map
        dataBillForMonth.forEach(v => {
            if (v.type === EnumBill.Out && v.category) {  // v.type === EnumBill.Out: 只统计支出的
                let prevAmount = mapCategoryAmount.get(v.category);
                prevAmount ? mapCategoryAmount.set(v.category, prevAmount + v.amount) : mapCategoryAmount.set(v.category, v.amount);
            }
        });

        // Map => Array
        let arrayCategoryAmount: Array<object> = [];
        mapCategoryAmount.forEach((v, k) => {
            const c = props.dataCategory.find(vc => vc.id === k);  // 为了分类(category)的名称(name)
            c && arrayCategoryAmount.push({
                key: c.id,
                category: c.name,
                amount: v
            });
        });
        return arrayCategoryAmount
    }

    return (
        <div>
            <Divider style={{ padding: "0 80px" }}>{ConstMonthsText[props.monthIndex]}账单分类统计</Divider>
            <Table
                pagination={false}
                scroll={{ y: 340 }}
                columns={outCategoryColumns}
                dataSource={checkOutCategoryData()}
            />
        </div>
    );
};

export default OutCategories;
