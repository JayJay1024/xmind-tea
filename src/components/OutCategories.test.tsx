import React from 'react';
import { RenderResult, render, act, screen } from '@testing-library/react';
import '../matchMedia.tsx';

import { EnumBill, EnumMonth, TypeBill, TypeCategory } from '../types';

import OutCategories from './OutCategories';

describe('<OutCategories />', () => {
    const fakeDataCategory: TypeCategory[] = [{
        id: 'cid_1111',
        name: 'cname_1111',
        type: EnumBill.In,
    }, {
        id: 'cid_2222',
        name: 'cname_2222',
        type: EnumBill.Out,
    }, {
        id: 'cid_3333',
        name: 'cname_3333',
        type: EnumBill.Out,
    }];
    const fakeDataBill: TypeBill[] = [{
        time: new Date('2048-12-12 11:10:20'),
        type: EnumBill.In,  // 收入，将不计算
        category: 'cid_1111',
        amount: 10,
    }, {
        time: new Date('2048-12-12 11:10:20'),
        type: EnumBill.Out,
        category: 'cid_2222',
        amount: 10,
    }, {
        time: new Date('2048-12-12 12:10:20'),
        type: EnumBill.Out,
        category: 'cid_2222',
        amount: 20,
    }, {
        time: new Date('2048-12-12 13:10:20'),
        type: EnumBill.Out,
        category: 'cid_3333',
        amount: 60,
    }, {
        time: new Date('2048-12-12 14:10:20'),
        type: EnumBill.Out,
        category: 'cid_3333',
        amount: 40,
    }];

    test('January no data', () => {
        // 根据 fakeDataBill ，事实只有 12 月有数据显示
        act(() => {
            render(<OutCategories dataBill={fakeDataBill} dataCategory={fakeDataCategory} monthIndex={EnumMonth.Jan} />);
        });
        expect(screen.queryByText('cname_1111')).toBeNull();
        expect(screen.queryByText('cname_2222')).toBeNull();
        expect(screen.queryByText('cname_3333')).toBeNull();
    });

    test('no income data', () => {
        act(() => {
            render(<OutCategories dataBill={fakeDataBill} dataCategory={fakeDataCategory} monthIndex={EnumMonth.Dec} />);
        });
        expect(screen.queryByText('cname_1111')).toBeNull();
        expect(screen.queryByText('cname_2222')).not.toBeNull();
        expect(screen.queryByText('cname_3333')).not.toBeNull();
    });

    test('calculate correctly', () => {
        act(() => {
            render(<OutCategories dataBill={fakeDataBill} dataCategory={fakeDataCategory} monthIndex={EnumMonth.Dec} />);
        });
        expect(screen.queryByText('30')).not.toBeNull();
        expect(screen.queryByText('100')).not.toBeNull();
    });

    test('should match snapshot', () => {
        let renderResult: any = null;
        act(() => {
            renderResult = render(<OutCategories dataBill={fakeDataBill} dataCategory={fakeDataCategory} monthIndex={EnumMonth.Dec} />);
        });
        expect((renderResult as RenderResult).asFragment()).toMatchSnapshot();
    });
});
