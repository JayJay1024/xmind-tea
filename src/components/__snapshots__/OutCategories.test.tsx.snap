// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OutCategories /> should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="ant-divider ant-divider-horizontal ant-divider-with-text ant-divider-with-text-center"
      role="separator"
      style="padding: 0px 80px;"
    >
      <span
        class="ant-divider-inner-text"
      >
        12月账单分类统计
      </span>
    </div>
    <div
      class="ant-table-wrapper"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table ant-table-fixed-header"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-header"
                style="overflow: hidden;"
              >
                <table
                  style="table-layout: fixed; visibility: hidden;"
                >
                  <colgroup />
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        账单类型
                      </th>
                      <th
                        class="ant-table-cell ant-table-column-sort ant-table-column-has-sorters"
                        style="text-align: center;"
                      >
                        <div
                          class="ant-table-column-sorters-with-tooltip"
                        >
                          <div
                            class="ant-table-column-sorters"
                          >
                            <span>
                              支出金额统计(￥)
                            </span>
                            <span
                              class="ant-table-column-sorter ant-table-column-sorter-full"
                            >
                              <span
                                class="ant-table-column-sorter-inner"
                              >
                                <span
                                  aria-label="caret-up"
                                  class="anticon anticon-caret-up ant-table-column-sorter-up"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    class=""
                                    data-icon="caret-up"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                    />
                                  </svg>
                                </span>
                                <span
                                  aria-label="caret-down"
                                  class="anticon anticon-caret-down ant-table-column-sorter-down active"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    class=""
                                    data-icon="caret-down"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                    />
                                  </svg>
                                </span>
                              </span>
                            </span>
                          </div>
                        </div>
                      </th>
                    </tr>
                  </thead>
                </table>
              </div>
              <div
                class="ant-table-body"
                style="overflow-y: scroll; max-height: 340px;"
              >
                <table
                  style="table-layout: fixed;"
                >
                  <colgroup />
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      aria-hidden="true"
                      class="ant-table-measure-row"
                      style="height: 0px;"
                    >
                      <td
                        style="padding: 0px; border: 0px; height: 0px;"
                      />
                      <td
                        style="padding: 0px; border: 0px; height: 0px;"
                      />
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="cid_3333"
                    >
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        cname_3333
                      </td>
                      <td
                        class="ant-table-cell ant-table-column-sort"
                        style="text-align: center;"
                      >
                        100
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="cid_2222"
                    >
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        cname_2222
                      </td>
                      <td
                        class="ant-table-cell ant-table-column-sort"
                        style="text-align: center;"
                      >
                        30
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
